src/models/geograph.py

```diff
@@ -614,21 +669,317 @@ def get_graph_components(
+        Class for the connected components of a GeoGraph.
+
+        This class can load a graph from only a list of components, in which
+        case it will not have a dataframe and the `has_df` class attribute will
+        be False, or it can load from both a list of components and a dataframe.
+        In the latter case, edges with the polygon distance can be added between
+        each pair of nodes (where each node corresponds to a connected component
+        in the original graph).
+
+        WARNING: using `add_distance_edges=True` is incredibly slow for anything
+        other than the smallest habitats.
+
+        Args:
+            components_list (List[set]): A list of sets, where each set contains
+            the node indices making up a component from the original graph.
+            df (Optional[gpd.GeoDataFrame], optional): A GeoDataFrame in which
+            each row contains a polygon that represents a component from a
+            GeoGraph. This is optional, and if not passed the graph will be
+            created with no edges. Defaults to None.
+            add_distance_edges (bool, optional): Boolean that determines whether
+            to add edges between every pair of nodes, with the distance between
+            their corresponding polygons as an edge attribute. Defaults to False.
+        """
+        # pylint: disable=super-init-not-called
+        self.has_df: bool = True
+        self.graph: nx.Graph = nx.Graph()
+        self.components_list: List[set] = components_list
+        if df is not None:
+            self.has_distance_edges: bool = add_distance_edges
+            self.df = self._load_from_dataframe(df, add_distance_edges)
+        else:
+            self.has_df = False
+            self.has_distance_edges = False
+            self.graph = nx.empty_graph(len(self.components_list))
+        self.metrics: Dict[str, Metric] = {}
+
+    def _load_from_dataframe(
+        self,
+        df: gpd.GeoDataFrame,
+        tolerance: float = 0.0,
+    ) -> gpd.GeoDataFrame:
+        """
+        Load graph from dataframe.
+
+        If `self.has_distance_edges` is True, then this will load a complete graph
+        and add the distance between each pair of polygons as an edge attribute.
+        Otherwise, it will load an empty graph.
+
+        In both cases, information about the polygons will be stored as node
+        attributes.
+
+        Args:
+            df (gpd.GeoDataFrame): GeoDataFrame containing polygon objects.
+            tolerance (float, optional): Redundant argument to ensure interface
+            is consistent with parent class. Defaults to 0.0.
+
+        Returns:
+            gpd.GeoDataFrame: A processed GeoDataFrame containing polygon objects.
+        """
+        # Reset index to ensure consistent indices
+        df = df.reset_index(drop=True)
+        # Using this list and iterating through it is slightly faster than
+        # iterating through df due to the dataframe overhead
```

Comment:
Did you benchmark with `df.geometry.values` (which gives you a geometry array object, that has references to the underlying C objects? How does this compare? 
(Feel free to address this after the report and just open an issue - I think before the report submission performance is not priority)

Reply:
Yeah I was thinking about this, I'll test it out before I merge the PR.

Reply:
@Croydon-Brixton I did some tests and it turns out making the list is slower than just getting the `.values`, but accessing a random element in the list is ~25x faster than accessing a random element in the geometry array (which is built on top of numpy I think). So we can stick to lists, this post might explain why they are better for accessing elements: https://stackoverflow.com/questions/29281680/numpy-individual-element-access-slower-than-for-lists